docs/TypeInference.md

```diff
@@ -0,0 +1,218 @@
+    Instead they are obtained entirely via a side effect operation - a call to the database.
+    This means that from the point of view of the type system, a table stream is just a list of **homogeneous** records.
+    In other words, record fields or columns that are never accessed never need to be typed.
+
+    Additionally if we assume transformations are given correct types, type inference can determine the columns and column types that the user expects from the database.
+    Storage can then implicitly filter out any types that do not match the user's expectation.
+
+### Typing flux functions using lists of records
+
+The goal of this 2nd iteration is **not** to assign an accurate type to every Flux transformation.
+As it stands, it is not possible to accurately describe certain Flux transformations with a type signature.
+Instead we should give every transformation a type signature of the form:
+
+```
+forall ['r, 's] where 'r:Rec, 's:Rec (tables: ['r], ...) -> ['s]
+```
+
+In words, every Flux transformation is a function which takes in a list of record kinds and returns a list of record kinds.
+This gives us parity with the current system while allowing us to iteratively assign more accurate and descriptive types, giving us the ability to catch more errors at compile time.
+
+Below I attempt to assign accurate types to common Flux transformations.
+
+* from
+
+    ```
+    forall ['r, 's] where 's: StrRec (org: str, bucket: str) -> [ {_time: time | _start: time | _stop: time | _value: 'r | 's} ]
+    ```
+    This polymorphic type signature says that `from` is a function that returns a list of records.
+    These records have fields `_time`, `_start`, `_stop`, and `_value`.
+    The `_value` field can be of any type.
+    All other fields represented by the row variable `'s` must be of type _string_.
+    Note the last part corresponds to the tag values of a series.
+
+    An example of a program that will fail to compile is:
+    ```
+    from(bucket: "telegraf")
+        |> range(start: -1h)
+        |> filter(fn: (r) => r.host == 1) // error: {host: int | 't} is not of kind StrRec
+    ```
+    The reason for failed compilation will be clear soon, but basically `range` and `filter` are just pass through transformations.
+    They don't extend or modify the types of their inputs.
+    As a result `filter` expects to receive the same list type that `from` returns and so it expects `r.host` to be of type _string_.
+    This contradicts the usage of `r.host` as an _int_.
+
+* range
+
+    ```
+    forall ['r] where 'r:Rec (start: time, stop: time, tables: [ {_start: time | _stop: time | 'r} ]) -> [ {_start: time | _stop: time | 'r} ]
+    ```
+    Let's focus on the `tables` parameter.
+    `range` takes a `tables` parameter that is a list of records that must have fields `_start` and `_stop` and returns the same list type.
```

Comment:
Technically `range` can accept data without `_start` and `_stop` columns and will add them if they do not exist. If my memory serves it is still possible to describe this using record extensions since our flavor will not allow duplicate fields. So it's implied they are update if they exist and added otherwise.

Taking a crack at describing this in the type language:

```
forall ['r] where 'r:Rec (start: time, stop: time, tables: [ { 'r} ]) -> [ {_start: time | _stop: time | 'r} ]
```

Reply:
Unfortunately record operations which either extend or update based on the presence of the field cannot be correctly type inferred in the proposed system. In particular we are unable to infer a principal type in certain situations. 

An example that I did not come up with:
```
f = (r) => if true then {r with x:1} else {x: 2}
```

If record extension includes update, then the type of `r` could be either `{}` or possibly `{x:int | {}}`.

As a result, our operations should either modify an already present field, or extend a record with a new field.

Fortunately I don't think this is too much of a restriction. In the case of `range` all it's doing is performing a filter on time. Hence it makes sense for `_start` and `_stop` to always be required.